### PR TITLE
perf(config): Migrate configSchema from zod to valibot (experimental)

### PR DESCRIPTION
**Pull request by yamadashy**
## Summary

Experimental migration of `src/config/configSchema.ts` from zod to valibot to evaluate cold-start impact as a simpler alternative to #1484 (which defers zod via async loading + a separate `configDefaults.ts`).

Draft PR — not for merge as-is. Kept for comparison with #1484.

## Benchmark Results

`hyperfine --warmup 3 --runs 20` packing a small 2-file target directory, 3 A/B rounds each:

| Scenario | main (zod v4) | valibot | Diff |
| --- | --- | --- | --- |
| No config file, R1 | 156.3 ± 3.2 ms | 145.4 ± 1.4 ms | **-10.9 ms (-7.0%)** |
| No config file, R2 | 155.8 ± 2.8 ms | 145.9 ± 1.8 ms | -9.9 ms (-6.4%) |
| No config file, R3 | 156.1 ± 2.2 ms | 146.8 ± 1.6 ms | -9.3 ms (-6.0%) |
| With config file, R1 | 160.9 ± 1.7 ms | 152.5 ± 4.0 ms | -8.4 ms (-5.2%) |
| With config file, R2 | 160.8 ± 2.3 ms | 151.3 ± 1.4 ms | -9.5 ms (-5.9%) |
| With config file, R3 | 161.7 ± 1.7 ms | 152.3 ± 1.7 ms | -9.4 ms (-5.8%) |

Average improvement: **~10 ms (~6%)** cold start, consistent with and without a config file.

Below the -15 ms / -50 ms thresholds used to decide between "adopt and close #1484" vs. "consult user" vs. "drop". Posting as-is for discussion.

## Trade-offs vs. #1484

- **Pro**: no `configDefaults.ts` split, no async API break in `loadFileConfig` / `mergeConfigs`, no speculative preload. Simpler end state.
- **Pro**: removes dependency on zod v4's auto-unwrap of ESM Module namespaces — `configLoad.ts` now unwraps `.default` explicitly, which works identically across libraries.
- **Con**: only ~10 ms faster. #1484 reports a larger win.
- **Con**: introduces a second validation library (valibot) while MCP tools still use zod. Full removal of zod would require migrating the MCP tool definitions too (out of scope here).

## API Surface

- `src/index.ts` re-exports `RepomixConfigFile as RepomixConfig` and `defineConfig` — structural shape preserved; user-facing `RepomixConfig` type is unchanged.
- Internal: `rethrowValidationErrorIfZodError` renamed to `rethrowValidationErrorIfSchemaError`. Duck-types both `ZodError` and `ValiError` so the helper stays library-agnostic.
- `repomixOutputStyleSchema` and other schema symbols are not part of the public export surface.

## Notable Implementation Detail

zod v4 silently auto-unwraps the `default` key when handed an ESM Module namespace object; valibot does not. `configLoad.ts` was relying on this implicitly for jiti-loaded `.ts` / `.mjs` configs. This PR makes the unwrap explicit:

```ts
const imported = await deps.jitiImport(pathToFileURL(filePath).href);
config =
  imported && typeof imported === 'object' && 'default' in imported
    ? (imported as { default: unknown }).default
    : imported;
```

This removes a subtle coupling to zod v4 behaviour, regardless of which library we end up keeping.

## Checklist

- [x] `npm run test` — 1115/1115 pass
- [x] `npm run lint` — clean (2 pre-existing warnings unrelated to this change)

## Related

- #1484 — async deferral approach (comparison baseline)

### CODE REVIEW

**Comment by yamadashy** [comment]
@gemini review

File: src/shared/errorHandle.ts [hunk 1]
```diff
@@ -115,20 +115,38 @@ const isRepomixError = (error: unknown): error is RepomixError => {
...
+    !('issues' in error) ||
+    !Array.isArray((error as { issues: unknown[] }).issues)
   ) {
-    const issues = (error as { issues: Array<{ path: string[]; message: string }> }).issues;
-    const zodErrorText = issues.map((err) => `[${err.path.join('.')}] ${err.message}`).join('\n  ');
-    throw new RepomixConfigValidationError(
-      `${message}\n\n  ${zodErrorText}\n\n  Please check the config file and try again.`,
-    );
+    return;
   }
```
[the remark this answers is not in the document]
Comment (yamadashy) [reply src/shared/errorHandle.ts:132]: Addressed in `1e8a5e5e` — switched to duck-typing so worker-serialized errors are recognized. Added a dedicated test in `tests/shared/errorHandle.test.ts`. 🤖

File: tests/config/configSchema.test.ts [hunk 2]
```diff
@@ -137,17 +137,17 @@ describe('configSchema', () => {
...
       const invalidConfig = {};
-      expect(() => repomixConfigDefaultSchema.parse(invalidConfig)).toThrow();
+      expect(() => v.parse(repomixConfigDefaultSchema, invalidConfig)).toThrow();
     });
 
     it('should provide helpful error for missing required fields', () => {
       const invalidConfig = {};
-      expect(() => repomixConfigDefaultSchema.parse(invalidConfig)).toThrow(/expected object/i);
+      expect(() => v.parse(repomixConfigDefaultSchema, invalidConfig)).toThrow(/expected|invalid/i);
     });
```
[the remark this answers is not in the document]
Comment (yamadashy) [reply tests/config/configSchema.test.ts:158]: Addressed in `1e8a5e5e` — replaced the broad regex with `toThrow(v.ValiError)` + a targeted `/invalid (type|key)/i` on `issues[0].message`. 🤖

File: website/client/src/public/schemas/latest/schema.json [hunk 3]
```diff
@@ -76,8 +77,7 @@
         },
         "splitOutput": {
           "type": "integer",
-          "minimum": 1,
-          "maximum": 9007199254740991
+          "minimum": 1
         },
```
[the remark this answers is not in the document]
Comment (yamadashy) [reply website/client/src/public/schemas/latest/schema.json:81]: Addressed in `1e8a5e5e` — added `v.maxValue(Number.MAX_SAFE_INTEGER)` to the `splitOutput` pipe so the generated schema keeps editor-hint parity with the previous zod output. 🤖

File: src/config/configSchema.ts [hunk 4]
```diff
@@ -14,151 +14,149 @@ export const defaultFilePathMap: Record<RepomixOutputStyle, string> = {
...
+  security: v.optional(
+    v.object({
+      enableSecurityCheck: v.optional(v.boolean()),
+    }),
+  ),
+  tokenCount: v.optional(
+    v.object({
+      encoding: v.optional(v.string()),
+    }),
+  ),
```
[the remark this answers is not in the document]
Comment (yamadashy) [reply src/config/configSchema.ts:74]: No action needed — this is the same layering that existed on `main` (base used `z.string().optional()` while the default schema used `z.enum(TOKEN_ENCODINGS).default(...)`). Not introduced by this migration, and tightening the base schema is a separate decision we can take in a follow-up. 🤖

File: src/config/configSchema.ts [hunk 5]
```diff
@@ -14,151 +14,149 @@ export const defaultFilePathMap: Record<RepomixOutputStyle, string> = {
...
+  v.object({
+    output: v.optional(
+      v.object({
+        stdout: v.optional(v.boolean()),
+      }),
+    ),
+    skillGenerate: v.optional(v.union([v.string(), v.boolean()])),
   }),
-);
+]);
```
[the remark this answers is not in the document]
Comment (yamadashy) [reply src/config/configSchema.ts:139]: No action needed — `v.intersect` only runs at config parse time (once per CLI invocation), so the delta vs. spread-composition is negligible next to the ~10ms cold-start win this PR targets. Keeping the intersect form also makes the layering between base / file / cli / default explicit. 🤖

File: src/config/configSchema.ts [hunk 6]
```diff
@@ -14,151 +14,149 @@ export const defaultFilePathMap: Record<RepomixOutputStyle, string> = {
...
-    }),
-  );
+export const repomixConfigMergedSchema = v.intersect([
+  repomixConfigDefaultSchema,
+  repomixConfigFileSchema,
+  repomixConfigCliSchema,
+  v.object({
+    cwd: v.string(),
+  }),
+]);
```
[the remark this answers is not in the document]
Comment (yamadashy) [reply src/config/configSchema.ts:153]: No action needed — `repomixConfigMergedSchema` intentionally layers default (required + defaults) over file (all optional) over cli. Spreading entries would have the later objects silently downgrade the required-with-default fields to optional, which changes semantics. Keeping the intersect form. 🤖

**Comment by yamadashy** [comment]
Thanks @coderabbitai / @gemini-code-assist / @claude for the reviews. Round-up of what landed in `1e8a5e5e`:

**Addressed:**
- Duck-type guard in `rethrowValidationErrorIfSchemaError` (CodeRabbit — errorHandle:132)
- Tightened test assertion to `toThrow(v.ValiError)` + targeted message pattern (CodeRabbit — configSchema.test:151, Claude — same point)
- Restored `splitOutput` upper bound to preserve schema parity (CodeRabbit — schema.json:81)
- Path-segment fallback returns `''` for unknown object shapes instead of `[object Object]` (CodeRabbit — errorHandle:134-148)
- Tightened ESM unwrap in `configLoad.ts`: only unwrap when `.default` is itself an object, preserving CJS configs that legitimately export a `default` key (CodeRabbit — configLoad:153-160)
- Stripped empty `required: []` arrays from generated schemas for output parity with the previous zod output (Claude — schema noise)
- Added `tests/shared/errorHandle.test.ts` covering Zod + Valibot + worker-serialized error paths (Claude — coverage gap)

**Skipped with reason:**
- `tokenCount.encoding` base-vs-default tightening (CodeRabbit — configSchema:74): pre-existing layering on `main` (`z.string().optional()` + `z.enum(...).default(...)`). Not introduced by this migration — separate decision.
- Numeric `integer()` / `minValue()` in base schema (CodeRabbit nitpick — configSchema:17-74): same story, pre-existing.
- Replace `v.intersect` with object-composition for `CliSchema` / `MergedSchema` (Gemini — configSchema:139, 149): `MergedSchema` intentionally layers default-with-defaults over file-as-optional; flattening via spread would silently downgrade required-with-default fields. Parse-time cost is negligible next to the cold-start target.
- Extra `v.looseObject` comment on `addAdditionalPropertiesFalse` (CodeRabbit nitpick — generateSchema:16-32): the existing comment already documents the "why" and the function is local to a single script.

**Open for discussion:**
- Dual runtime libs (CodeRabbit outside-diff — package.json) — depends on whether we commit to migrating the MCP tools too, or fall back to #1484. Noted as a trade-off in the PR description.
- Cherry-pick the `.default` unwrap to `main` as a standalone bugfix (Claude) — worth doing if this PR is dropped.
- `configLoad.ts` at 262 lines slightly over the 250-line guideline (Claude) — not splitting yet; open to doing so in follow-up.

**Comment by yamadashy** [comment]
Decisions on the open items:

1. **Dual runtime libs**: accepted. The MCP TypeScript SDK requires zod for tool input schemas, so we can't remove zod without giving up MCP. valibot only loads on the CLI config path, zod only loads when the MCP server starts — the ~10ms cold-start win is preserved for the default CLI path. Install-size impact is ~10KB (valibot).

2. **Cherry-pick `.default` unwrap to main**: not needed — we're merging this PR, so the fix ships with it.

3. **`configLoad.ts` 262 lines**: leaving as-is for this PR to keep the scope tight. 12 lines over the guideline is minor; happy to split in a follow-up if anyone wants to pick that up.

**Comment by yamadashy** [comment]
Responding to @claude's follow-up review on `1e8a5e5`:

**Addressed:**
- ESM unwrap edge cases now have test coverage — added in `193e4dd7`. `configLoad.integration.test.ts` now exercises both the `{ default: configObject }` unwrap branch and the pathological CJS `{ default: 'plain', ...rest }` pass-through branch via mocked `jitiImport`.

**Not addressed (previously decided):**
- Base vs. default numeric constraints: the ask is valid but this layering is pre-existing from the zod version — scope for a follow-up schema-design PR, not this migration.
- `configLoad.ts` 265 lines: deferring to a follow-up split as previously discussed.

Thanks to @coderabbitai, @gemini-code-assist, @devin-ai-integration, and @claude for the reviews — all actionable items from this round are resolved.

**Comment by yamadashy** [comment]
Addressing @claude's final-assessment finding in `d83224a7`:

**Numeric constraint behavioral coverage** — added 8 focused rejection cases in `tests/config/configSchema.test.ts` under `repomixConfigDefaultSchema > numeric constraint enforcement`:

- `maxFileSize`: rejects non-integer (`1.5`) and below-minimum (`0`)
- `topFilesLength`: rejects negative (`-1`)
- `splitOutput`: rejects below 1 (`0`), non-integer (`1.5`), and above `Number.MAX_SAFE_INTEGER`
- `git.sortByChangesMaxCommits` / `git.includeLogsCount`: reject below 1

This proves the `v.pipe(v.number(), v.integer(), v.minValue(...), v.maxValue(...))` pipes are behaviorally equivalent to the previous `z.number().int().min(...)` chains, not just structurally.

Follow-up: mirrored `v.maxValue(Number.MAX_SAFE_INTEGER)` onto `repomixConfigDefaultSchema.splitOutput` so the upper bound is enforced no matter which schema the caller parses against (the earlier edit only touched the base schema). Generated JSON schema is unchanged since it's built from the file/base schema.

**Comment by yamadashy** [comment]
Responding to @claude's two post-`d83224a` reviews:

**Claude review #1 (pre-`d83224a` context):**
1. `splitOutput` asymmetry between base and default schemas — already addressed in `d83224a7` (reviewer hadn't seen it yet; the maxValue on the default schema landed in that commit).
2. Added a WHY comment near `repomixConfigMergedSchema` in `91dcd2cc` so future readers know the `v.intersect` choice is load-bearing (spreading would silently demote required-with-default fields).
3. `errorHandle.ts:142` uncovered branch — leaving as-is. The branch is a defensive fallback for a malformed ValiError shape that isn't emitted in practice. Adding a fabricated path-item test to hit it would go against the project convention of not adding error handling / tests for scenarios that can't happen.

**Claude review #2:**
- `errorHandle.ts:142` — same as above.
- `defaultAction.ts:350` `throw error` — also a defensive re-throw; reached only if `v.parse()` somehow throws a non-schema error, which isn't a real scenario. Leaving uncovered.

Codecov's patch coverage sits at 91.67% and the 3 missing lines are all of this defensive variety. Not worth synthetic tests.

Thanks for the review rounds — I'll treat this PR as ready pending the merge-vs-drop decision against #1484.

**Comment by yamadashy** [comment]
Responding to @claude's latest review (in `7c2e8791`):

**Addressed:**
1. **`interopDefault: true` + manual unwrap redundancy** — dropped `interopDefault: true` from the jiti config. The explicit ESM namespace unwrap at the call site already handles every module-format case (.ts / .mts / .js / .mjs / .cjs), all integration tests still pass.
2. **Empty path segments producing `..`** — added a `.filter((s) => s !== '')` before the `.join('.')` so a malformed ValiError path item (object without `key`) drops out cleanly instead of yielding `[output..style]`. Covered by a new unit test.

**Skipped with reason:**
- **ESM unwrap worth cherry-picking to main** — orthogonal to this PR's code. If this PR merges, the fix ships with it; if dropped, happy to revisit as a standalone PR (previously captured in the PR discussion).
- **`maxFileSize` lacks constraint in base schema** — pre-existing layering from the zod era (`z.number()` in base + `z.number().int().min(1)` in default), not introduced by this migration.
- **Lockfile / `scripts/memory/tsconfig.json`** — npm artifacts and a necessary CI fix for TS6's `moduleResolution: "node"` deprecation that was blocking the Memory Benchmark job.
- **`normalizeObjectNode` untested / no `satisfies` type-compat check / uncovered defensive branches** — script-local helper and defensive code paths. Skipping per the "don't test / don't guard against scenarios that can't happen" project convention.

Error-handler compliments noted — thanks!

**Comment by yamadashy** [comment]
Responding to @claude's latest review (`e2a63eed`):

**Addressed:**
- **`v.intersect` CLI-only field preservation** — added a targeted test in `tests/config/configSchema.test.ts` that parses a merged config with `output.stdout: true` and `skillGenerate: 'my-skill'` and asserts both survive the intersect. Confirmed valibot merges per-schema outputs correctly.

**Skipped with reason:**
- **`splitOutput` new `maxValue(Number.MAX_SAFE_INTEGER)`** — intentional, added in response to an earlier CodeRabbit thread about restoring schema.json parity with the zod-generated output. The commit body of `1e8a5e5e` documents this.
- **Dual runtime deps** — accepted trade-off (MCP SDK requires zod), documented in the PR description.
- **Cherry-pick ESM unwrap** — ships with this PR if merged; happy to separate if this branch is dropped.
- **`configLoad.ts` 268 lines** — deferred to a follow-up split (previously agreed).
- **`encoding` asymmetry** — pre-existing layering from the zod era, not introduced by this migration.
- **Minor test gaps (base-schema splitOutput, path-less issues, multi-issue formatting, `default: null`)** — defensive scenarios, not core behavioral coverage; the intersect stdout test above is the one real correctness gap worth pinning.
- **Pin `@valibot/to-json-schema` + snapshot test** — reasonable follow-up but out of scope for this PR; the current patch is minimal enough to re-derive if the converter output shifts.

**Comment by yamadashy** [comment]
Responding to @claude's 6-agent review (`9c779459`):

**Addressed:**
1. **ESM unwrap edge case with object-valued `default`** — added a comment explicitly documenting the limitation (CJS module of shape `{ default: {...}, otherKey: ... }` would be truncated). No known user hits this since `RepomixConfig` has no `default` field. Preferred the comment over a `Symbol.toStringTag === 'Module'` check because jiti's TS interop for `.ts` / `.mts` returns `{ default: ... }` wrappers that lack the Module tag, so that guard would break the common path.
2. **`normalizeObjectNode` mutates in-place** — added a one-line comment noting the in-place mutation above the function.

**Skipped (acknowledgment / observational):**
- v.intersect comment being "load-bearing" — thanks, keeping it.
- Dual-library runtime cost — documented trade-off in the PR description, accepted given MCP SDK's zod requirement.
- Base/default constraint asymmetry — pre-existing from zod era, out of migration scope.
- Test file sizes / `defaultAction.ts` over 250 lines — observational, previously deferred.
- Error-handler duck-typing / `() => []` default factory praise — noted, thanks.

**Comment by yamadashy** [comment]
Responding to @claude's latest 6-agent review (`744bb7fc`):

**Addressed:**
- **`repomixConfigCliSchema` field preservation in isolation** — added a test that parses `{ output: { filePath, stdout }, skillGenerate }` directly against `repomixConfigCliSchema` and asserts all three fields survive the intersect. This pins the pre-merge step (`buildCliConfig` in `defaultAction.ts:347`) separately from the existing `MergedSchema` test.

No other actionable items — the remaining review content is positive verification across the six analysis passes. Thanks for the thorough read.

**Comment by yamadashy** [comment]
Responding to @claude's latest review (`53ddb5ac`):

**Addressed:**
- **`normalizeObjectNode` has no test coverage** — extracted the helper to its own module (`website/client/scripts/normalizeJsonSchema.ts`) and added `tests/website/normalizeJsonSchema.test.ts` with 7 cases covering `additionalProperties` add/preserve, empty `required` stripping, recursion through `properties` / `anyOf` / `items`, and non-object / bare-object passthrough. Re-ran `website-generate-schema` to confirm the output is byte-identical after the refactor.

**Skipped with reason:**
- **`v.intersect` redundant base-schema evaluation** — measurable cost is sub-ms per `mergeConfigs` call, and duplicating the base shape into a flat `repomixConfigMergedSchema` to avoid it would duplicate ~30 fields and introduce a drift risk. Keeping the intersect; the performance impact is dominated by jiti / IO elsewhere on the cold path.
- **`configLoad.ts` 270 lines** — deferred to a follow-up split (previously agreed in multiple rounds).
- **Module-level `v.parse` of `defaultConfig` runs at every cold start** — not quite: `configSchema.ts` is only reached via `configLoad.ts` imported by `defaultAction.ts`, and `defaultAction.ts` is dynamically imported from `cliRun.ts` after the `--help` / `--version` yargs short-circuit. So the eager parse runs only when we're about to do real work anyway.

**Comment by yamadashy** [comment]
Responding to @claude's latest review (`85b0c7a7`):

**Addressed:**
- **ESM unwrap ambiguous case** — added an integration test in `tests/config/configLoad.integration.test.ts` that pins the documented limitation (object `default` + sibling keys → siblings are discarded). Non-issue for RepomixConfig today, but frozen so behavior can't drift silently.
- **Empty path in error formatting** — `rethrowValidationErrorIfSchemaError` now emits just `message` when a schema issue has no path, instead of `[] message`. Covered by a new test case in `tests/shared/errorHandle.test.ts`.

**Skipped with reason:**
- **Dual validation libraries ship** — accepted trade-off (MCP SDK requires zod), previously documented.
- **Base vs. default constraint mismatch** — pre-existing from the zod era, out of migration scope. Re-raised across several review rounds; maintaining the prior decision.
- **Min-boundary acceptance tests** — the existing rejection tests already establish the fencepost (`maxFileSize: 0` rejects, and `validConfig` tests use valid boundary values). Adding explicit `maxFileSize: 1` acceptance tests would be redundant with the valid-config tests already in the suite.
- ESM unwrap / error-handler / normalizeObjectNode praise items — noted, thanks.

**Comment by yamadashy** [comment]
Responding to @claude's latest review (`b27bebb1`):

**Addressed:**
- **Non-integer rejection parity** — added cases for `topFilesLength`, `git.sortByChangesMaxCommits`, and `git.includeLogsCount` so every `v.pipe(..., v.integer(), ...)` numeric field has matching rejection coverage alongside the min-bound tests that were already in place.

**Skipped with reason:**
- Strategic / architecture observations (dual-library path, `defaultConfig` init fragility, `normalizeJsonSchema` maintenance debt) — all previously discussed and documented; no new actionable items.
- 2 uncovered defensive lines (`throw error` fallthrough paths) — unreachable in normal operation, consistent with the prior skip decisions.
